### PR TITLE
Add ECDsa span-based methods

### PR DESCRIPTION
**Pull request by stephentoub**
cc: @bartonjs 
Closes https://github.com/dotnet/corefx/issues/22615

### CODE REVIEW

File: src/Common/src/System/Security/Cryptography/ECDsaCng.SignVerify.cs [hunk 1]
```diff
@@ -51,6 +51,14 @@ public override byte[] SignHash(byte[] hash)
             }
         }
 
+        public override unsafe bool TrySignHash(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten)
```
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/ECDsaCng.SignVerify.cs]: What here (and at VerifyHash) requires the unsafe keyword?

Comment (stephentoub) [inline src/Common/src/System/Security/Cryptography/ECDsaCng.SignVerify.cs]: The `null` argument to TrySignHash is a `void*`.

**Comment by stephentoub** [comment]
Thanks for reviewing, @bartonjs.

**Comment by ektrah** [comment]
@stephentoub - Quick question: The [code currently pins](https://github.com/stephentoub/corefx/blob/b06ddbabcbde91fc3839ad023534c225653df0da/src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.RSA.cs#L47-L66) spans for P/Invoke, e.g.:

```csharp
private static unsafe int RsaEncryptPkcs(
    SafeSecKeyRefHandle publicKey,
    ReadOnlySpan<byte> pbData,
    int cbData,
    out SafeCFDataHandle pEncryptedOut,
    out SafeCFErrorHandle pErrorOut)
{
    fixed (byte* pbDataPtr = &pbData.DangerousGetPinnableReference())
    {
        return RsaEncryptPkcs(publicKey, pbDataPtr, cbData, out pEncryptedOut, out pErrorOut);
    }
}
```

Is that required or could the P/Invoke method defined to accept a `ref byte`? E.g.:

```csharp
private static int RsaEncryptPkcs(
    SafeSecKeyRefHandle publicKey,
    ReadOnlySpan<byte> pbData,
    out SafeCFDataHandle pEncryptedOut,
    out SafeCFErrorHandle pErrorOut)
{
    return RsaEncryptPkcs(publicKey, pbData.DangerousGetPinnableReference(), pbData.Length, out pEncryptedOut, out pErrorOut);
}

[DllImport(Libraries.AppleCryptoNative, EntryPoint = "AppleCryptoNative_RsaEncryptPkcs")]
private static extern int RsaEncryptPkcs(
    SafeSecKeyRefHandle publicKey,
    ref byte pbData,
    int cbData,
    out SafeCFDataHandle pEncryptedOut,
    out SafeCFErrorHandle pErrorOut);
```

I'm wondering because manual pinning is usually not required when passing, for example, a field by reference:

```csharp
class A { public int B; }

[DllImport(...)]
static extern void Foo(ref int value);

static void Main()
{
    var a = new A();
    Foo(ref a.B);
}
```

I would have assumed that the CLR automatically pins the object in either case.

**Comment by stephentoub** [comment]
> Is that required or could the P/Invoke method defined to accept a ref byte?

That's a good question.  I actually hadn't considered doing that, but I don't see why it wouldn't work.

**Pull request merged**